### PR TITLE
Eq/Ord instances for Variant

### PR DESCRIPTION
**Pull request by natefaubion**

### CODE REVIEW

File: src/Data/Variant/Internal.purs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  variantTags ∷ LProxy rl → L.List String
+
+instance variantTagsNil ∷ VariantTags R.Nil where
+  variantTags _ = L.Nil
+
+instance variantTagsCons ∷ (VariantTags rs, IsSymbol sym) ⇒ VariantTags (R.Cons sym a rs) where
+  variantTags _ = L.Cons (reflectSymbol (SProxy ∷ SProxy sym)) (variantTags (LProxy ∷ LProxy rs))
+
+lookupEq
+  ∷ ∀ r
```
Comment (LiamGoodacre) [inline src/Data/Variant/Internal.purs]: Unsused `r`.

**Comment by LiamGoodacre** [comment]
This looks good :+1:

I think it'd be worth trying to have a general version of the two lookup functions. Perhaps combining with variantTags to reduce boilerplate - all the functions are basically the same.

I've just been off on a tangent thinking about a crazy version of this with deciding Leibniz equality of reified keys and a record of well-typed binary functions instead of the two lists.  I got **very** carried away.

Here's a rough sketch if you care to delve into it 😈 - In which `BinaryVariantSub r o ro` replaces a row `r = { x :: x, y :: y }` with `ro = { x :: x -> x -> o, y :: y -> y -> o }`:

```purescript
class BinaryVariantSub r o ro <= BinaryVariant (r :: # Type) o | r -> o where
  binaryVariant :: forall lk rk lt.
    -- keys lk and rk are reified variant case keys
    -- key lk associated with type lt
    RowCons lk lt _ r =>
    -- key rk is in r
    RowCons rk _ _ r =>
    -- proxy for inference
    RProxy r ->
    -- decide equality of keys or result `o`
    (SProxy lk -> SProxy rk -> Either (Leibniz (SProxy lk) (SProxy rk)) o)
    -- record of binary functions to `o`
    Record ro ->
    -- lookup values of lk and rk, if keys are the same then values are the same type
    (Leibniz (SProxy lk) (SProxy rk) -> {lhs :: lt, rhs :: lt}) ->
    -- result
    o
```

You'd need to have a way of reifying a variant key with the extra constraint that the key is a valid option of the variant.

Complexity wise, I think the only real thing this would have over your current version is the near constant lookup time of the functions from a record instead of walking through the lists you pass around.  Or in your case, maybe `StrMap (VariantCase -> VariantCase -> o)`.

Anyway.  You're probably better off with staying with `String` and `VariantCase` - it's a lot shorter (less to mess up/maintain).

I might have a play around with this at the weekend and see how far we can take it.

**Comment by natefaubion** [comment]
There are a couple reasons I went with the two lists:

* Building a record or StrMap as a dictionary might be faster for dispatch (my original approach did this), but I think it could be slower to build. If the compiler was better about sharing and lifting instances, I'd probably go that route. You could maybe do it through unsafe mutation, since it should actually be safe. It's probably worth a benchmark in any case. The list approach is still just linear, as is normal pattern matching with large ADTs.
* By building a separate list of tags, it can be shared when you instantiate `Ord`. Otherwise you end up with a lot of duplicate `IsSymbol` dictionaries all over the place, and this makes a huge difference with large variants. Again, I wish the compiler was better about sharing these.

**Comment by natefaubion** [comment]
I split out the dispatch function. I think this is a good initial implementation given my goals. Maybe when we flesh out a records library, we can revisit it if there are more efficient operations to work with. Thanks @LiamGoodacre.

**Comment by LiamGoodacre** [comment]
Yeah looks great :+1:

**Pull request merged**